From 1375fc3a0f8ad8216183e6f3626a1ee0ce199efd Mon Sep 17 00:00:00 2001
From: Himmat Lal <quic_hlal@quicinc.com>
Date: Fri, 21 Jul 2023 00:47:36 +0530
Subject: [PATCH] net: Add chan to freq formula for new added channel

Add channel to frequency conversion formula for
new added channels in ieee80211_channel_to_frequency() API.
This API is invoked by the driver when the frame is passed on
to cfg80211 layer as a part of scan

Signed-off-by: Himmat Lal <quic_hlal@quicinc.com>
Change-Id: I3e323635d04ed300f735ffa8cbe736ee6a528e74
---
 net/wireless/util.c | 18 ++++++++++++++----
 1 file changed, 13 insertions(+), 3 deletions(-)

diff --git a/net/wireless/util.c b/net/wireless/util.c
index 151407522111..010064fe5f72 100644
--- a/net/wireless/util.c
+++ b/net/wireless/util.c
@@ -86,10 +86,18 @@ u32 ieee80211_channel_to_freq_khz(int chan, enum nl80211_band band)
 			return MHZ_TO_KHZ(2407 + chan * 5);
 		else if (chan >= 221 && chan <= 224)
 			return MHZ_TO_KHZ(2477 + (chan - 221) * 5);
-		else if (chan > 200)
+		else if (chan > 200 && chan <= 224)
 			return MHZ_TO_KHZ(2399 + (chan - 200) * 5);
+		else if ((chan > 190 && chan < 200) || chan > 224) {
+			if ((chan % 2) == 0)
+				return MHZ_TO_KHZ(2399 + (((chan - 190)*5)/2));
+			else
+				return MHZ_TO_KHZ(2399 + (((chan - 190)*5)/2) +1);
+		}
 		break;
	case NL80211_BAND_5GHZ:
+		if ((chan > 1 && chan < 31) || (chan >= 182 && chan <= 184))
+			return MHZ_TO_KHZ(5000 + chan * 5);
 		if (chan >= 182 && chan <= 196)
 			return MHZ_TO_KHZ(4000 + chan * 5);
 		else
@@ -103,7 +111,7 @@ u32 ieee80211_channel_to_freq_khz(int chan, enum nl80211_band band)
 			return MHZ_TO_KHZ(5950 + chan * 5);
 		break;
 	case NL80211_BAND_60GHZ:
-		if (chan < 7)
+		if (chan < 5)
 			return MHZ_TO_KHZ(56160 + chan * 2160);
 		break;
 	default:
@@ -121,7 +129,9 @@ int ieee80211_freq_khz_to_channel(u32 freq)
 	/* see 802.11 17.3.8.3.2 and Annex J */
 	if (freq == 2484)
 		return 14;
-	else if (freq < 2484)
+	else if (freq < 2412 || (freq > 2492 && freq <= 2499))
+		return (190 + (((freq - 2399) * 2)/5));
+	else if (freq < 2484 && freq >= 2412)
 		return (freq - 2407) / 5;
 	else if (freq >= 4910 && freq <= 4980)
 		return (freq - 4000) / 5;
-- 
2.25.1

